File: dotnet/samples/Demos-Before/BookingRestaurant/BookingsPlugin.cs
// Copyright (c) Microsoft. All rights reserved.

using System.ComponentModel;
using Microsoft.Graph;
using Microsoft.Graph.Models;
using Microsoft.SemanticKernel;

namespace Plugins;

/// <summary>
/// Booking Plugin with specialized functions for booking a table at a restaurant using Microsoft Graph Bookings API.
/// </summary>
internal sealed class BookingsPlugin
{
    private readonly GraphServiceClient _graphClient;
    private readonly string _businessId;
    private readonly string _customerTimeZone;
    private readonly string _serviceId;

    private const int PostBufferMinutes = 10;
    private const int PreBufferMinutes = 5;

    internal BookingsPlugin(
        GraphServiceClient graphClient,
        string businessId,
        string serviceId,
        string customerTimeZone = "America/Chicago"
    )
    {
        this._graphClient = graphClient;
        this._businessId = businessId;
        this._serviceId = serviceId;
        this._customerTimeZone = customerTimeZone;
    }

    [KernelFunction("BookTable")]
    [Description("Books a new table at a restaurant")]
    public async Task<string> BookTableAsync(
        [Description("Name of the restaurant")] string restaurant,
        [Description("The time in UTC")] DateTime dateTime,
        [Description("Number of people in your party")] int partySize,
        [Description("Customer name")] string customerName,
        [Description("Customer email")] string customerEmail,
        [Description("Customer phone number")] string customerPhone
    )
    {
        Console.WriteLine($"System > Do you want to book a table at {restaurant} on {dateTime} for {partySize} people?");
        Console.WriteLine("System > Please confirm by typing 'yes' or 'no'.");
        Console.Write("User > ");
        var response = Console.ReadLine()?.Trim();
        if (string.Equals(response, "yes", StringComparison.OrdinalIgnoreCase))
        {
            var requestBody = new BookingAppointment
            {
                OdataType = "#microsoft.graph.bookingAppointment",
                CustomerTimeZone = this._customerTimeZone,
                SmsNotificationsEnabled = false,
                EndDateTime = new DateTimeTimeZone
                {
                    OdataType = "#microsoft.graph.dateTimeTimeZone",
                    DateTime = dateTime.AddHours(2).ToString("o"),
                    TimeZone = "UTC",
                },
                IsLocationOnline = false,
                OptOutOfCustomerEmail = false,
                AnonymousJoinWebUrl = null,
                PostBuffer = TimeSpan.FromMinutes(PostBufferMinutes),
                PreBuffer = TimeSpan.FromMinutes(PreBufferMinutes),
                ServiceId = this._serviceId,
                ServiceLocation = new Location
                {
                    OdataType = "#microsoft.graph.location",
                    DisplayName = restaurant,
                },
                StartDateTime = new DateTimeTimeZone
                {
                    OdataType = "#microsoft.graph.dateTimeTimeZone",
                    DateTime = dateTime.ToString("o"),
                    TimeZone = "UTC",
                },
                MaximumAttendeesCount = partySize,
                FilledAttendeesCount = partySize,
                Customers =
                [
                    new BookingCustomerInformation
                    {
                        OdataType = "#microsoft.graph.bookingCustomerInformation",
                        Name = customerName,
                        EmailAddress = customerEmail,
                        Phone = customerPhone,
                        TimeZone = this._customerTimeZone,
                    },
                ],
                AdditionalData = new Dictionary<string, object>
                {
                    ["priceType@odata.type"] = "#microsoft.graph.bookingPriceType",
                    ["reminders@odata.type"] = "#Collection(microsoft.graph.bookingReminder)",
                    ["customers@odata.type"] = "#Collection(microsoft.graph.bookingCustomerInformation)"
                },
            };

            // list service IDs
            var services = await this._graphClient.Solutions.BookingBusinesses[this._businessId].Services.GetAsync();

            // To initialize your graphClient, see https://learn.microsoft.com/en-us/graph/sdks/create-client?from=snippets&tabs=csharp
            var result = await this._graphClient.Solutions.BookingBusinesses[this._businessId].Appointments.PostAsync(requestBody);

            return "Booking successful!";
        }

        return "Booking aborted by the user";
    }

    [KernelFunction]
    [Description("List reservations booking at a restaurant.")]
    public async Task<List<Appointment>> ListReservationsAsync()
    {
        // Print the booking details to the console
        var resultList = new List<Appointment>();
        var appointments = await this._graphClient.Solutions.BookingBusinesses[this._businessId].Appointments.GetAsync();

        foreach (var appointmentResponse in appointments?.Value!)
        {
            resultList.Add(new Appointment(appointmentResponse));
        }

        return resultList;
    }

    [KernelFunction]
    [Description("Cancels a reservation at a restaurant.")]
    public async Task<string> CancelReservationAsync(
        [Description("The appointment ID to cancel")] string appointmentId,
        [Description("Name of the restaurant")] string restaurant,
        [Description("The date of the reservation")] string date,
        [Description("The time of the reservation")] string time,
        [Description("Number of people in your party")] int partySize)
    {
        // Print the booking details to the console
        Console.ForegroundColor = ConsoleColor.DarkBlue;
        Console.WriteLine($"System > [Cancelling a reservation for {partySize} at {restaurant} on {date} at {time}]");
        Console.ResetColor();

        await this._graphClient.Solutions.BookingBusinesses[this._businessId].Appointments[appointmentId].DeleteAsync();

        return "Cancellation successful!";
    }
}

File: python/samples/demos/booking_restaurant/bookings_plugin/bookings_plugin.py
# Copyright (c) Microsoft. All rights reserved.

from datetime import datetime, timedelta
from typing import Annotated

from msgraph import GraphServiceClient
from msgraph.generated.models.booking_appointment import BookingAppointment
from msgraph.generated.models.booking_customer_information import BookingCustomerInformation
from msgraph.generated.models.date_time_time_zone import DateTimeTimeZone
from msgraph.generated.models.location import Location

from semantic_kernel.functions.kernel_function_decorator import kernel_function


class BookingsPlugin:
    """A plugin for booking tables at a restaurant."""

    def __init__(
        self,
        graph_client: GraphServiceClient,
        booking_business_id: str,
        booking_service_id: str,
        customer_timezone: str = "America/Chicago",
    ):
        """Initializes a new instance of the BookingsPlugin class.

        Args:
            graph_client (GraphServiceClient): The GraphServiceClient instance.
            booking_business_id (str): The ID of the booking business.
            service_id (str): The ID of the service.
            customer_timezone (str, optional): The timezone of the customer. Defaults to "America/Chicago".
        """
        self.graph_client = graph_client
        self.booking_business_id = booking_business_id
        self.booking_service_id = booking_service_id
        self.customer_timezone = customer_timezone

    @kernel_function(name="book_table", description="Book a table at a restaurant")
    async def book_table(
        self,
        restaurant: Annotated[str, "The name of the restaurant"],
        date_time: Annotated[str, "The time in UTC, formatted as an ISO datetime string, like 2024-09-15T19:00:00"],
        party_size: Annotated[int, "The number of people in the party"],
        customer_name: Annotated[str, "The name of the customer"],
        customer_email: Annotated[str, "The email of the customer"],
        customer_phone: Annotated[str, "The phone number of the customer"],
    ) -> Annotated[str, "The booking appointment ID"]:
        """Book a table at a restaurant.

        Args:
            restaurant (str): The name of the restaurant.
            date_time (datetime): The time in UTC.
            party_size (int): The number of people in the party.
            customer_name (str): The name of the customer.
            customer_email (str): The email of the customer.
            customer_phone (str): The phone number of the customer.

        Returns:
            str: The status of the booking.
        """
        print(f"System > Do you want to book a table at {restaurant} on {date_time} for {party_size} people?")
        print("System > Please confirm by typing 'yes' or 'no'.")
        confirmation = input("User:> ")
        if confirmation.lower() != "yes":
            return "Booking aborted by the user."
        request_body = BookingAppointment(
            odata_type="#microsoft.graph.bookingAppointment",
            customer_time_zone=self.customer_timezone,
            sms_notifications_enabled=False,
            start_date_time=DateTimeTimeZone(
                odata_type="#microsoft.graph.dateTimeTimeZone",
                date_time=date_time,
                time_zone="UTC",
            ),
            end_date_time=DateTimeTimeZone(
                odata_type="#microsoft.graph.dateTimeTimeZone",
                date_time=(datetime.fromisoformat(date_time) + timedelta(hours=2)).isoformat(),
                time_zone="UTC",
            ),
            is_location_online=False,
            opt_out_of_customer_email=False,
            anonymous_join_web_url=None,
            service_id=self.booking_service_id,
            service_location=Location(
                odata_type="#microsoft.graph.location",
                display_name=restaurant,
            ),
            maximum_attendees_count=party_size,
            filled_attendees_count=party_size,
            customers=[
                BookingCustomerInformation(
                    odata_type="#microsoft.graph.bookingCustomerInformation",
                    name=customer_name,
                    email_address=customer_email,
                    phone=customer_phone,
                    time_zone=self.customer_timezone,
                ),
            ],
            additional_data={
                "price_type@odata_type": "#microsoft.graph.bookingPriceType",
                "reminders@odata_type": "#Collection(microsoft.graph.bookingReminder)",
                "customers@odata_type": "#Collection(microsoft.graph.bookingCustomerInformation)",
            },
        )

        response = await self.graph_client.solutions.booking_businesses.by_booking_business_id(
            self.booking_business_id
        ).appointments.post(request_body)

        return f"Booking successful! Your reservation ID is {response.id}."

    @kernel_function(name="list_revervations", description="List all reservations")
    async def list_reservations(self) -> Annotated[str, "The list of reservations"]:
        """List the reservations for the booking business."""
        appointments = await self.graph_client.solutions.booking_businesses.by_booking_business_id(
            self.booking_business_id
        ).appointments.get()
        return "\n".join([
            f"{appointment.service_location.display_name} on {appointment.start_date_time.date_time} with id: {appointment.id}"  # noqa: E501
            for appointment in appointments.value
        ])

    @kernel_function(name="cancel_reservation", description="Cancel a reservation")
    async def cancel_reservation(
        self,
        reservation_id: Annotated[str, "The ID of the reservation"],
        restaurant: Annotated[str, "The name of the restaurant"],
        date: Annotated[str, "The date of the reservation"],
        time: Annotated[str, "The time of the reservation"],
        party_size: Annotated[int, "The number of people in the party"],
    ) -> Annotated[str, "The cancellation status of the reservation"]:
        """Cancel a reservation."""
        print(f"System > [Cancelling a reservation for {party_size} at {restaurant} on {date} at {time}]")

        _ = (
            await self.graph_client.solutions.booking_businesses.by_booking_business_id(self.booking_business_id)
            .appointments.by_booking_appointment_id(reservation_id)
            .delete()
        )
        return "Cancellation successful!"
